fuel-vm/src/tests/blockchain.rs

```diff
@@ -268,27 +270,188 @@ fn state_read_write() {
+    let target_contract = vec![
         op::log(RegId::ONE, RegId::ONE, RegId::ONE, RegId::ONE),
         op::ret(RegId::ONE),
         op::ret(RegId::ZERO), // Pad to make length uneven to test padding
     ];
 
-    let program: Witness = contract_code.into_iter().collect::<Vec<u8>>().into();
+    let bytes = target_contract.into_iter().collect::<Vec<u8>>();
+    let len = bytes.len().try_into().unwrap();
+    let offset = 0;
+    let program: Witness = bytes.into();
 
-    let contract = Contract::from(program.as_ref());
+    let receipts = ldc__load_len_of_target_contract(
+        &mut client,
+        rng,
+        salt,
+        offset,
+        len,
+        program.clone(),
+    );
+
+    if let Receipt::LogData { digest, .. } = receipts.get(0).expect("No receipt") {
+        let mut code = program.into_inner();
+        code.extend([0; 4]);
+        assert_eq!(digest, &Hasher::hash(&code), "Loaded code digest incorrect");
+    } else {
+        panic!("Script did not return a value");
+    }
+
+    if let Receipt::Log { ra, .. } = receipts.get(1).expect("No receipt") {
+        assert_eq!(*ra, 1, "Invalid log from loaded code");
+    } else {
+        panic!("Script did not return a value");
+    }
+}
+
+#[test]
+fn ldc__gas_cost_is_not_dependent_on_rC() {
+    let rng = &mut StdRng::seed_from_u64(2322u64);
+    let salt: Salt = rng.gen();
+
+    let mut client = MemoryClient::default();
+
+    let gas_costs = client.gas_costs();
+    let ldc_cost = gas_costs.ldc;
+    let ldc_dep_len = ldc_cost.dep_per_unit;
+    let noop_cost = gas_costs.noop;
+
+    let contract_size = 1000;
+    let offset = 0;
+    let starting_rC = 0;
+    let starting_gas_amount =
+        ldc__gas_cost_for_len(&mut client, rng, salt, contract_size, offset, starting_rC);
+
+    for i in 1..10 {
+        // Increase by ldc_dep_cost for each attempt
+        let len_diff = (i * ldc_dep_len) as u16;
+        let len = starting_rC + len_diff;
+        // The gas should go up 0 per ldc_dep_cost and 1 per noop_cost every 4
+        // bytes (noop is 4 bytes)
+        let cost_of_noops = len_diff as u64 / 4 * noop_cost;
+        let expected_gas_used = starting_gas_amount + cost_of_noops;
+
+        let actual_gas_used =
+            ldc__gas_cost_for_len(&mut client, rng, salt, contract_size, offset, len);
+        assert_eq!(actual_gas_used, expected_gas_used);
+    }
+}
+
+#[test]
+fn ldc__offset_affects_read_code() {
+    let rng = &mut StdRng::seed_from_u64(2322u64);
+    let salt: Salt = rng.gen();
+
+    let mut client = MemoryClient::default();
+
+    let gas_costs = client.gas_costs();
+    let noop_cost = gas_costs.noop;
+
+    let contract_size = 1000;
+    let offset = 0;
+    let len = 100;
+    let starting_gas_amount =
+        ldc__gas_cost_for_len(&mut client, rng, salt, contract_size, offset, len);
+
+    for i in 1..10 {
+        let offset = i * 4;
+        let expected_gas_used = starting_gas_amount - (i * noop_cost);
```

Comment:
Hmm. I just realized that this test shouldn't be passing--or at least I don't know why it's passing.

Since, the `len` is staying the same, each of the copied codes should be identical. Indeed, if I debug the code that's what I'm seeing. But for some reason execution of that code isn't identical.

In fact, locally, I've changed `contract_size` to be `25` so that the remaining bytes are empty. In that case I'd expect the test to pass, but it fails because each is 2 less than original, rather than just 1.

Reply:
Indeed the gas amount should be the same. What is the value of `noop_cost`? Maybe it is zero

Reply:
The bug was in the interpreter code. I was calling `copy_from_slice_zero_fill_noownerchecks` with the wrong values.